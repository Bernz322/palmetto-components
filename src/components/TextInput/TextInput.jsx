import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { v4 as uuid } from 'uuid';
import Cleave from 'cleave.js/react';
import * as InputMasks from './TextInputMasks';
import './TextInput.scss';
import FormLabel from '../FormLabel/FormLabel';

const getInputMask = (inputMask, availableInputMasks) => {
  if (typeof inputMask === 'string') {
    return availableInputMasks[inputMask];
  }

  return inputMask;
};

const Input = ({
  autoComplete,
  autoFocus,
  className,
  id,
  inputMask,
  isDisabled,
  isRequired,
  label,
  name,
  onBlur,
  onChange,
  onFocus,
  placeholder,
  type,
  value,
}) => {
  const handleChange = e => {
    onChange(e);
  };

  const handleFocus = e => {
    if (onFocus) onFocus(e);
    e.currentTarget.select(); // Selects input content allowing immediate edit. @TODO -- Confirm if desired functionality.
  };

  const handleBlur = e => {
    if (onBlur) onBlur(e);
  };

  const inputClasses = classNames(
    className,
    'input',
  );

  const inputId = id || uuid();

  const inputProps = {
    'aria-required': isRequired,
    'aria-label': label || name,
    autoComplete,
    autoFocus,
    className: inputClasses,
    disabled: isDisabled,
    id: inputId,
    name,
    onBlur: handleBlur,
    onChange: handleChange,
    onFocus: handleFocus,
    placeholder,
    type,
    value,
  };

  const labelProps = {
    isFieldRequired: isRequired,
    inputId,
    labelText: label,
  };

  return (
    <>
      {label && <FormLabel {...labelProps} />}
      {!inputMask ? (
        <input {...inputProps} />
      ) : (
        <Cleave {...inputProps} options={getInputMask(inputMask, InputMasks)} />
      )}
    </>
  );
};

Input.propTypes = {
  /**
   * The input's 'autocomplete' attribute
   */
  autoComplete: PropTypes.oneOf(['on', 'off']),
  /**
   * The input's 'autocomplete' attribute
   */
  autoFocus: PropTypes.bool,
  /**
   * Custom class to be added to standard input classes.
   */
  className: PropTypes.string,
  /**
   * The input's id attribute. We will use this to programmatically tie the input with its label.
   */
  id: PropTypes.string,
  /**
   * Pass a value to apply a mask to the input value.
   * Can be one of the existing present strings, or a custom object with options
   * For options object formats See https://github.com/nosir/cleave.js
   */
  inputMask: PropTypes.oneOfType([
    PropTypes.oneOf(['phone', 'creditCard']),
    PropTypes.object,
  ]),
  /**
   * The input's disabled attribute
   */
  isDisabled: PropTypes.bool,
  /**
   * Determines if input is required or not. (Label will have an asterisk if required)
   */
  isRequired: PropTypes.bool,
  /**
   * Value for HTML <label> tag
   */
  label: PropTypes.string,
  /**
   * The input's 'name' attribute
   */
  name: PropTypes.string,
  /**
   * Callback function to call on blur event.
   */
  onBlur: PropTypes.func,
  /**
   * Callback function to call on change event.
   */
  onChange: PropTypes.func.isRequired,
  /**
   * Callback function to call on focus event..
   */
  onFocus: PropTypes.func,
  /**
   * The input placeholder attribute
   */
  placeholder: PropTypes.string,
  /**
   * The input 'type' value. Defaults to type 'text'.
   */
  type: PropTypes.oneOf(['text', 'password', 'email', 'tel', 'url', 'search']),
  /**
   * The text value of the input. Required since our Input is a controlled component.
   */
  value: PropTypes.string.isRequired,
};

Input.defaultProps = {
  autoComplete: 'off',
  autoFocus: false,
  className: '',
  id: undefined,
  inputMask: undefined,
  isDisabled: false,
  isRequired: false,
  label: undefined,
  name: '',
  onBlur: undefined,
  onFocus: undefined,
  placeholder: '',
  type: 'text',
};

export default Input;
