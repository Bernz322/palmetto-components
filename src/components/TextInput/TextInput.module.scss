@import '../../styles/globals/form-controls.scss';

.textInput {
  composes: baseInput;
}
