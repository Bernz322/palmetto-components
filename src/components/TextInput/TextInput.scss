@import '../../styles/variables.scss';

/* @TODO -- replace any remaining harcoded values with variables or use global utility classes */
.Palmetto-TextInput {
  background-color: var(--color-base-white);
  border: 1px solid var(--color-font-grey-lighter);
  box-sizing: border-box;
  box-shadow: 0px 3px 0px rgba(0, 0, 0, 0.05);
  border-radius: var(--size-border-radius-md);
  padding: var(--size-spacing-sm);
  font-size: var(--size-font-base);
  color: var(--color-font-base);
  width: 100%;
  transition-duration: 300ms;
  transition-property: border, background-color;
  transition-timing-function: cubic-bezier(.2, .8, .4, 1);

  &::placeholder {
    opacity: 1;
    /* Override lower placeholder opacity in Firefox */
    color: var(--color-font-grey-light);
  }

  &:focus {
    border-color: var(--color-base-grey-600);
    outline: none;
  }

  &:disabled {
    color: var(--color-font-grey-light);
    background-color: var(--color-base-grey-50);

    &:hover {
      cursor: not-allowed;
    }
  }

  &.error {
    background-color: var(--color-base-red-50);
    border-color: var(--color-base-red-200);

    &:focus {
      background-color: var(--color-base-white);
      border-color: var(--color-base-grey-600);
      outline: none;
    }
  }
}
