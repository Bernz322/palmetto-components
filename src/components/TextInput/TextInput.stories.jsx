import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { action } from '@storybook/addon-actions';
import TextInput from './TextInput';

export default {
  title: 'TextInput',
  component: TextInput,
};

const StatefulInput = props => {
  const { initialValue } = props;
  const [value, setValue] = useState(initialValue);

  const handleChange = e => {
    action('onChange')(e);
    setValue(e.target.value || e.target.rawValue);
  };

  const handleFocus = e => {
    action('onFocus')(e);
  };

  const handleBlur = e => {
    action('onBlur')(e);
  };

  return (
    <TextInput
      value={value}
      onChange={handleChange}
      onFocus={handleFocus}
      onBlur={handleBlur}
      {...props} // eslint-disable-line react/jsx-props-no-spreading
    />
  );
};

StatefulInput.propTypes = {
  initialValue: PropTypes.string.isRequired,
};

export const All = () => {
  const basicInputValue = 'Hello World!';
  const disabledInputValue = 'I am disabled';
  const placeholderInputValue = '';
  const withLabelInputValue = 'With a label';
  const requiredInputValue = '';
  const autoFocusedInputValue = '';
  const withInputMaskValue = '';

  return (
    <div style={{ width: '400px' }}>
      <div style={{ marginBottom: '1rem' }}>
        <StatefulInput initialValue={basicInputValue} />
      </div>
      <div style={{ marginBottom: '1rem' }}>
        <StatefulInput initialValue={disabledInputValue} isDisabled />
      </div>
      <div style={{ marginBottom: '1rem' }}>
        <StatefulInput initialValue={withLabelInputValue} label="Name" />
      </div>
      <div style={{ marginBottom: '1rem' }}>
        <StatefulInput initialValue={requiredInputValue} label="Required Input" isRequired />
      </div>
      <div style={{ marginBottom: '1rem' }}>
        <StatefulInput
          initialValue={placeholderInputValue}
          label="With placeholder"
          placeholder="I am a placeholder"
        />
      </div>
      <div style={{ marginBottom: '1rem' }}>
        <StatefulInput
          initialValue={autoFocusedInputValue}
          label="Autofocused"
          placeholder="I am autofocused on page load"
          autoFocus
        />
      </div>
      <div style={{ marginBottom: '1rem' }}>
        <StatefulInput
          initialValue={withInputMaskValue}
          label="With phone input mask"
          placeholder="I have a phone number format"
          type="tel"
          inputMask="phone"
        />
      </div>
    </div>
  );
};
