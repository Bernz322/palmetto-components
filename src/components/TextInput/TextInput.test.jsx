import React from 'react';
import {
  render,
  fireEvent,
  screen,
  waitFor,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import TextInput from './TextInput';

describe('TextInput', () => {
  /**
   * Props Validation
   */
  test('Throws error if required prop "value" is not supplied to component', () => {
    console.error = jest.fn(); // eslint-disable-line no-console
    render(<TextInput onChange={() => null} />);
    expect(console.error).toHaveBeenCalledTimes(1); // eslint-disable-line no-console
    expect(console.error.mock.calls[0][0]) // eslint-disable-line no-console
      .toContain('Failed prop type: The prop `value`');
  });

  test('Throws error if required prop "onChange" is not supplied to component', () => {
    console.error = jest.fn(); // eslint-disable-line no-console
    render(<TextInput value="hello" />);
    expect(console.error).toHaveBeenCalledTimes(1); // eslint-disable-line no-console
    expect(console.error.mock.calls[0][0]) // eslint-disable-line no-console
      .toContain('Failed prop type: The prop `onChange`');
  });

  test('Throws an error if "type" prop is anything other than allowed values', () => {
    console.error = jest.fn(); // eslint-disable-line no-console
    render(<TextInput onChange={() => null} value="hello" type="notOnTheList" />);
    expect(console.error).toHaveBeenCalledTimes(1); // eslint-disable-line no-console
    expect(console.error.mock.calls[0][0]) // eslint-disable-line no-console
      .toContain('Failed prop type: Invalid prop `type`');
  });

  /**
   * Component Functionality
   */
  test('onChange event fires callback function', () => {
    const onChangeMock = jest.fn(() => null);

    render(
      <TextInput
        name="firstName"
        id="firstName"
        label="first name"
        value="hello"
        onChange={onChangeMock}
      />,
    );
    const inputElement = screen.getByDisplayValue('hello');

    fireEvent.change(inputElement, { target: { value: 'good bye' } });
    expect(onChangeMock).toHaveBeenCalledTimes(1);
  });

  test('Input value is updated properly when upper state changes', () => {
    let value = 'hello';
    const onChangeMock = jest.fn(event => { value = event.target.value; });
    const { rerender } = render(
      <TextInput
        name="firstName"
        id="firstName"
        label="first name"
        value={value}
        onChange={onChangeMock}
      />,
    );

    const inputElement = screen.getByDisplayValue('hello');
    expect(inputElement.value).toBe('hello');

    fireEvent.change(inputElement, { target: { value: 'good bye' } });
    expect(onChangeMock).toHaveBeenCalledTimes(1);

    rerender(
      <TextInput
        name="firstName"
        id="firstName"
        label="first name"
        value={value}
        onChange={onChangeMock}
      />,
    );
    expect(inputElement.value).toBe('good bye');
  });

  test('Input fires onFocus callback', () => {
    const mockedOnFocus = jest.fn();
    render(<TextInput value="hello" onChange={() => null} onFocus={mockedOnFocus} />);
    const inputElement = screen.getByDisplayValue('hello');
    fireEvent.focus(inputElement);
    expect(mockedOnFocus).toBeCalledTimes(1);
  });

  test('Input fires onBlur callback', () => {
    const mockedOnBlur = jest.fn();
    render(<TextInput value="hello" onChange={() => null} onBlur={mockedOnBlur} />);
    const inputElement = screen.getByDisplayValue('hello');
    fireEvent.focus(inputElement);
    fireEvent.blur(inputElement);
    expect(mockedOnBlur).toBeCalledTimes(1);
  });

  test('Input autofocuses if "autoFocus" prop is set to true', () => {
    render(<TextInput value="hello" onChange={() => null} autoFocus />);
    const inputElement = screen.getByDisplayValue('hello');
    expect(document.activeElement).toEqual(inputElement);
  });

  /**
   * Snapshots
   * NOTE: Even though 'id' is not a required prop for this component we must provide it for the test
   * to pass since the component will otherwise generate a uuid which means the snapshot is different every time.
   */
  test('Matches the snapshot for component with default props', () => {
    const { asFragment } = render(<TextInput value="" onChange={() => null} id="myId" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('Matches the snapshot for a disabled input', () => {
    const { asFragment } = render(<TextInput value="" onChange={() => null} id="myId" isDisabled />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('Matches the snapshot for a required input', () => {
    const { asFragment } = render(<TextInput value="" onChange={() => null} id="myId" isRequired />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('Matches the snapshot for an input with an error but no validation message', () => {
    const { asFragment } = render(<TextInput value="" onChange={() => null} id="myId" error />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('Matches the snapshot for an input with an error and a validation message', () => {
    const { asFragment } = render(<TextInput value="" onChange={() => null} id="myId" error="You silly goose" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
