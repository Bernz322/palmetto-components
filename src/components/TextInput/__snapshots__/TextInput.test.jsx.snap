// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Matches the snapshot for a disabled input 1`] = `
<DocumentFragment>
  <div>
    <input
      aria-invalid="false"
      aria-label=""
      aria-required="false"
      autocomplete="off"
      class="TextInput"
      disabled=""
      id="myId"
      name=""
      placeholder=""
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TextInput Matches the snapshot for a required input 1`] = `
<DocumentFragment>
  <div>
    <input
      aria-invalid="false"
      aria-label=""
      aria-required="true"
      autocomplete="off"
      class="TextInput"
      id="myId"
      name=""
      placeholder=""
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TextInput Matches the snapshot for an input with an error and a validation message 1`] = `
<DocumentFragment>
  <div>
    <input
      aria-invalid="true"
      aria-label=""
      aria-required="false"
      autocomplete="off"
      class="TextInput error"
      id="myId"
      name=""
      placeholder=""
      type="text"
      value=""
    />
    <div
      class="font-color-danger font-size-sm"
    >
      You silly goose
    </div>
  </div>
</DocumentFragment>
`;

exports[`TextInput Matches the snapshot for an input with an error but no validation message 1`] = `
<DocumentFragment>
  <div>
    <input
      aria-invalid="true"
      aria-label=""
      aria-required="false"
      autocomplete="off"
      class="TextInput error"
      id="myId"
      name=""
      placeholder=""
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TextInput Matches the snapshot for component with default props 1`] = `
<DocumentFragment>
  <div>
    <input
      aria-invalid="false"
      aria-label=""
      aria-required="false"
      autocomplete="off"
      class="TextInput"
      id="myId"
      name=""
      placeholder=""
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
